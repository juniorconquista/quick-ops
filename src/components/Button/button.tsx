import React, { memo } from 'react';
import styled from 'styled-components';
import { variant, VariantArgs, color, space } from 'styled-system';

import { ButtonProps, ButtonStyledProps } from './button.types';

/* TODO This component was created in the local repository as it has not yet been 
implemented in lib Quarks UI, however in the future it can be changed to be used directly from lib. */

const variants = variant({
    prop: 'styling',
    variants: {
        base: {
            height: '36px',
            borderRadius: '6px',
            color: '#6b7480 !important',
            border: '1px solid',
            borderColor: '#6b7480',
            p: '0px 20px',
            display: 'flex',
            alignItems: 'center',
            bg: 'white.default',
            cursor: 'pointer',
            transition: 'all .25s ease',
            fontFamily: 'Nunito Sans, sans-serif',
            svg: {
                height: '12px',
                width: 'auto',
                mr: '5px',
                fill: '#6b7480',
            },
            '&:hover, &:focus': {
                boxShadow: '0 0 2px 3px #6b748090',
            },
        },
        get success() {
            return {
                ...this.base,
                bg: 'success.default',
                borderColor: 'success.default',
                color: '#FFFFFF',
                '&:hover, &:focus': {
                    boxShadow: (props: {
                        colors: { success: { default: string } };
                    }) => `0 0 2px 3px ${props.colors.success.default}90`,
                },
                svg: {
                    // @ts-ignore
                    ...this.base.svg,
                    fill: '#FFFFFF',
                },
            };
        },
        get danger() {
            return {
                ...this.base,
                bg: 'danger.default',
                borderColor: 'danger.default',
                color: '#FFFFFF',
                '&:hover, &:focus': {
                    boxShadow: (props: {
                        colors: { danger: { default: string } };
                    }) => `0 0 2px 3px ${props.colors.danger.default}90`,
                },
                svg: {
                    // @ts-ignore
                    ...this.base.svg,
                    fill: '#FFFFFF',
                },
            };
        },
    },
} as VariantArgs);

const Style: React.FC<ButtonStyledProps> = styled.button<ButtonStyledProps>`
    ${variants}
    ${color}
    ${space}
    ${(props) =>
        props.outline &&
        `
        background-color: transparent;
    `}
`;

const Button: React.FC<ButtonProps> = ({
    styling = 'base',
    text,
    icon,
    ...props
}) => (
    <Style
        styling={styling}
        color={props.outline ? `${styling}.default` : '#FFFFFF'}
        {...props}
    >
        {icon}
        {text}
    </Style>
);

export default memo(Button);
