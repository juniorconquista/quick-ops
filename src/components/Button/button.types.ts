import { HTMLProps, ReactNode } from 'react';
import { ColorProps, SpaceProps } from 'styled-system';

export type ButtonStyledProps = ColorProps &
    SpaceProps &
    HTMLProps<HTMLButtonElement> & {
        /** Each styling variant has unique properties that are inherited through the type informed */
        styling?: 'base' | 'success' | 'danger';
        outline?: boolean;
    };

export interface ButtonProps extends ButtonStyledProps {
    text?: string;
    icon?: ReactNode;
    children?: ReactNode;
}
