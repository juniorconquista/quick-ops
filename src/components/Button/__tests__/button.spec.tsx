import React from 'react';
import renderer from 'react-test-renderer';
import { cleanup, render } from '@testing-library/react';
import ThemeProvider from '../../../utils/ThemeProviderTest';

import 'jest-styled-components';
import '@testing-library/jest-dom/extend-expect';

import Button from '../button';

describe('Button component', () => {
    afterEach(() => {
        jest.clearAllMocks();
        cleanup();
    });
    test('basic snapshot', () => {
        const container = render(
            <ThemeProvider>
                <Button styling="base" />
            </ThemeProvider>,
        );
        expect(container.asFragment()).toMatchSnapshot();
    });

    it('should inherit the styling properties of the success variant', () => {
        const button = renderer
            .create(
                <ThemeProvider>
                    <Button styling="success" />
                </ThemeProvider>,
            )
            .toJSON();
        expect(button).toHaveStyleRule('border-color', '#51c1c3');
        expect(button).toHaveStyleRule('color', '#FFFFFF');
        expect(button).toHaveStyleRule('background-color', '#51c1c3');
        expect(button).toHaveStyleRule('box-shadow', '0 0 2px 3px #51c1c390', {
            modifier: '&:hover',
        });
        expect(button).toHaveStyleRule('box-shadow', '0 0 2px 3px #51c1c390', {
            modifier: '&:focus',
        });
        expect(button).toHaveStyleRule('fill', '#FFFFFF', {
            modifier: 'svg',
        });
    });

    it('should inherit the styling properties of the danger variant', () => {
        const button = renderer
            .create(
                <ThemeProvider>
                    <Button styling="danger" />
                </ThemeProvider>,
            )
            .toJSON();
        expect(button).toHaveStyleRule('border-color', '#ff4138');
        expect(button).toHaveStyleRule('color', '#FFFFFF');
        expect(button).toHaveStyleRule('background-color', '#ff4138');
        expect(button).toHaveStyleRule('box-shadow', '0 0 2px 3px #ff413890', {
            modifier: '&:hover',
        });
        expect(button).toHaveStyleRule('box-shadow', '0 0 2px 3px #ff413890', {
            modifier: '&:focus',
        });
        expect(button).toHaveStyleRule('fill', '#FFFFFF', {
            modifier: 'svg',
        });
    });
});
