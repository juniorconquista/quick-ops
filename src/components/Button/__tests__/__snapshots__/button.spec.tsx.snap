// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button component basic snapshot 1`] = `
<DocumentFragment>
  .c0 {
  height: 36px;
  border-radius: 6px;
  color: #6b7480 !important;
  border: 1px solid;
  border-color: #6b7480;
  padding: 0px 20px;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  background-color: #25262a;
  cursor: pointer;
  -webkit-transition: boxShadow .25s ease;
  transition: boxShadow .25s ease;
  font-family: Nunito Sans,sans-serif;
  color: #FFFFFF;
}

.c0 svg {
  height: 15px;
  width: 15px;
  margin-right: 5px;
  fill: #6b7480;
}

.c0:hover,
.c0:focus {
  box-shadow: 0 0 2px 3px #6b748090;
}

<button
    class="c0"
    color="#FFFFFF"
  />
</DocumentFragment>
`;
